src/clarity-icons/clarity-icons-element.ts

```diff
@@ -55,25 +59,99 @@ const setIconSize = function(element: any, size: string) {
 };
 
 ClarityIconElement.prototype.connectedCallback = function() {
+    // One thing to note here is that the attributeChangedCallback method is called for every attribute first
+    // before this connectedCallback method called only once when the custom element is inserted into the DOM.
+    // So we could check whether the attribute values really changed or not.
+    // If not, we don't need to execute the same codes again.
     const host = this as HTMLElement;
 
     if (host.hasAttribute("shape")) {
-        generateIcon(host, host.getAttribute("shape"));
+        const shapeAttrValue = host.getAttribute("shape").split(/\s/)[0];
+
+        if (shapeAttrValue === this.currentShapeAttrVal) {
+            return;
+        }
+
+        this.currentShapeAttrVal = shapeAttrValue;
+
+        if (ClarityIconsApi.instance.has(this.currentShapeAttrVal)) {
+            this.currentShapeTemplate = ClarityIconsApi.instance.get(this.currentShapeAttrVal);
+        } else {
+            console.error(`'${this.currentShapeAttrVal}' is not found in the Clarity Icons set.`);
+            host.innerHTML = ClarityIconsApi.instance.get("error");
+            return;
+        }
+    }
+
+    if (host.hasAttribute("title")) {
+        const titleAttrValue = host.getAttribute("title");
+
+        if (this.currentTitleAttrVal !== titleAttrValue) {
+            this.currentTitleAttrVal = titleAttrValue;
+        }
+
+        if (!this.currentShapeAttrVal) {
+            return;
+        }
     }
 
     if (host.hasAttribute("size")) {
-        setIconSize(host, host.getAttribute("size"));
+        const sizeAttrValue = host.getAttribute("title");
+
+        if (this.currentSizeAttrVal !== sizeAttrValue) {
+            this.currentSizeAttrVal = sizeAttrValue;
+            setIconSize(host, host.getAttribute("size"));
+        }
     }
+
+    this.injectTemplate(host);
 };
 
 ClarityIconElement.prototype.attributeChangedCallback = function(attributeName: string, oldValue: string,
                                                                  newValue: string) {
     const host = this as HTMLElement;
 
+    if (!this.initialized) {
+        this.clrIconUniqId = "_clr_icon_" + clrIconId;
```

Comment:
Why is this set in the `attributeChangedCallback`? This is initialization, it clearly belongs to `connectedCallback`. You wouldn't even need the `this.initialized` property and test.

Reply:
What I figured out was that the `attributeChangedCallback()` is invoked before `connectedCallback()`. So I had to initialize the properties here with `this.initialized` guard. The `connectedCallback()` method gets called only once when it's actually inserted into the document tree. Of course, if any of the attributes are changed after the element is inserted into the dom, `attributeChangedCallback()` will be called after `connectedCallback()` as well.

Reply:
Ideally, I needed to initialize them in the constructor. But `this` in the constructor and the ones in any of the prototype methods of `ClarityIconElement` were pointing to different object. But I figured out why that was so and made changes accordingly to fix that issue.